portal-web/test/functional/com/liferay/portalweb/tests/enduseree/formsworkflow/metrics/WorkflowMetrics.testcase

```diff
@@ -304,6 +304,103 @@ definition {
 			velocityValueTotal = "0");
 	}
 
+	@description = "LRQA-53639 - Automate LPS-100675"
+	test VerifyFilterTopStepPerformance {
+
+		property portal.acceptance = "true";
+
+		var key_testSLADefinitionHours = "0001";
+		var key_testSLADefinitionHoursCheck = "00:01";
+		var key_testSLADefinitionDescription = "SLA Description";
+		var key_testSLADefinitionEnd = "Process Ends: Approved";
+		var key_testSLADefinitionName = "SLA Name";
+		var key_testSLADefinitionStart = "Process Begins";
+
+		WorkflowMetrics.goToWorkflowSLASettings(workflowProcessName = "Single Approver");
+
+		Click(locator1 = "WorkflowMetricsSLA#ADD_NEW_SLA_BUTTON");
+
+		WorkflowMetrics.fillSLAFormHours(
+			slaDefinitionDescription = "${key_testSLADefinitionDescription}",
+			slaDefinitionEnd = "${key_testSLADefinitionEnd}",
+			slaDefinitionHours = "${key_testSLADefinitionHours}",
+			slaDefinitionName = "${key_testSLADefinitionName}",
+			slaDefinitionStart = "${key_testSLADefinitionStart}");
+
+		Button.clickSave();
+
+		Alert.viewSuccessMessage();
+
+		WorkflowMetrics.goToCreatedSLAForm(workflowSLAName = "${key_testSLADefinitionName}");
+
+		AssertTextEquals(
+			locator1 = "WorkflowMetricsSLA#SLA_DEFINITION_NAME",
+			value1 = "${key_testSLADefinitionName}");
+
+		AssertTextEquals(
+			locator1 = "WorkflowMetricsSLA#SLA_DEFINITION_DESCRIPTION",
+			value1 = "${key_testSLADefinitionDescription}");
+
+		AssertTextEquals(
+			locator1 = "WorkflowMetricsSLA#SLA_STEPS_START",
+			value1 = "${key_testSLADefinitionStart}");
+
+		AssertTextEquals(
+			locator1 = "WorkflowMetricsSLA#SLA_STEPS_STOP",
+			value1 = "${key_testSLADefinitionEnd}");
+
+		AssertTextEquals(
+			locator1 = "WorkflowMetricsSLA#SLA_DEFINITION_HOURS",
+			value1 = "${key_testSLADefinitionHoursCheck}");
+
+		ProductMenu.gotoPortlet(
+			category = "Configuration",
+			panel = "Site Administration",
+			portlet = "Workflow");
+
+		Workflow.configureWorkflow(
+			workflowDefinition = "Single Approver",
+			workflowResourceValue = "Blogs Entry");
+
+		JSONBlog.addEntry(
+			entryContent = "Blogs Entry1 Content",
+			entryTitle = "Blogs Entry Title");
+
+		UserBar.gotoDropdownItem(dropdownItem = "My Workflow Tasks");
+
+		Workflow.assignToMeTaskByActions(
+			workflowAssetTitle = "Blogs Entry Title",
+			workflowAssetType = "Blogs Entry",
+			workflowTask = "Review");
+
+		Pause(locator1 = "5000");
+
+		Workflow.approveTaskByActions(
+			workflowAssetTitle = "Blogs Entry Title",
+			workflowAssetType = "Blogs Entry",
+			workflowTask = "Review");
+
+		ProductMenu.gotoPortlet(
+			category = "Workflow",
+			panel = "Control Panel",
+			portlet = "Metrics");
+
+		WorkflowMetrics.goToWorkflowProcessMetrics(workflowProcessName = "Single Approver");
+
+		Click(locator1 = "WorkflowMetrics#METRICS_PERFORMANCE_TAB");
+
+		WorkflowMetrics.assertDefaultChartRangesAndUnits();
+
+		WorkflowMetrics.assertChartTotalValuePerformanceByStep(slaBreachedValue = "1 (100%)");
```

Comment:
Hi Michael,
In order to this number to change from 0 (0%) to 1 (100%), the instance needs to completed after 1 minute as I said in my previous comment. 

In Workflow, the SLA engine updates after 1 minute, but we can't know for sure when the value will change from 0 to 1. We have 2 options to solve this problem:

1 - Add an explicit wait, like... wait 150 seconds and check if the number has changed. I don't link this approach because the time to update can be smaller, so we would have been spending a unnecessary time.

2 - Choose an element of the page that you know for sure that will change as well, an monitor that change. As soon as it changes, we can validate 1 (100%). One element that will have to change for sure, is the number of "Overdue" instances from Completed Items card. It will have to change from 0 to 1. So, as soon as it changes, we can validate the Performance by Step card.

I drafter a macro that waits for it:

	_macro waitForSLAtoRun {
		var count = "1";

		while (IsElementNotPresent(locator1 = "WorkflowMetrics#DASHBOARD_TOTAL_OVERDUE_ITEMS_VALUE") && ("${count}" != "15")) {

			echo("Waiting 10 seconds, until the SLA becomes overdue");

			Pause(locator1 = "10000");

			//Add a function here to refresh the page. We need to search for one.

			count = MathUtil.sum("${count}", "1");
		}

		if ("${count}" == "15") {
			echo("timeout exceeded");
		}	
	}_

So, before doing this assert, we would need to confirm that the page is indeed ready to be validated. 

Please let me know if any questions.

Reply:
**DASHBOARD_TOTAL_OVERDUE_ITEMS_VALUE**
_//span[contains(text(),'Overdue')]/ancestor::div[contains(@class, 'header')]/../div[contains(text(), '1')]_

WorkflowMetrics#DASHBOARD_TOTAL_OVERDUE_ITEMS_VALUE path (hardcoded with value 1. We would have to change that as well.

Reply:
So I could not find a way to do this without creating a path to look for an element with text in it. As far as I could tell "while" will not run using a locator and a value. Also, for some reason refresh did not want to work for this macro when I wrote it, but I could get around that by having it renavigate to the page. I cute down the time to be a max of 60 seconds before it times out.